tensorflow/compiler/mlir/tensorflow/transforms/convert_tf_control_flow_to_scf.cc

```diff
@@ -98,11 +103,86 @@ class ConvertIfRegionOp : public OpRewritePattern<IfRegionOp> {
   }
 };
 
+/// Convert the `tf.WhileRegion` op to the `scf.while` op iff the condition of
+/// the `tf.WhileRegion` op is a 0-D tensor of 1-bit signless integers.
+class ConvertWhileRegionOp : public OpRewritePattern<WhileRegionOp> {
+ public:
+  using OpRewritePattern<WhileRegionOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(WhileRegionOp op,
+                                PatternRewriter& rewriter) const override {
+    // Creates the `before` or `after` region of the `scf.while` op. Note that
+    // `tf_cond_or_body_region` is the `cond` or `body` region of the
+    // `tf.WhileRegion` op. `scf_before_or_after_region` is the `before` or
+    // `after` region of the new `scf.while` op.
+    auto createScfCondOrBody = [](Region& tf_cond_or_body_region,
+                                  Region& scf_before_or_after_region,
+                                  PatternRewriter& rewriter) {
+      Block* first_block_of_scf_before_or_after_region =
+          cloneRegionAndGetFirstBlock(tf_cond_or_body_region,
+                                      scf_before_or_after_region, rewriter);
```

Comment:
Cloning has non-trivial cost, can you look if you can use `rewriter.mergeBlocks()` to move the block instead?

Reply:
Let me know when this is addressed.

Reply:
@joker-eph, I tried doing this but due to the following assertion in the `rewriter.mergeBlocks()` function, it can be used in the `if` op lowering, but cannot be used in the `while` op lowering:
`assert(llvm::all_of(source->getPredecessors(), [dest](Block *succ) { return succ == dest; }) && "expected 'source' to have no predecessors or only 'dest'");`

What are your views?

Reply:
I'm a bit confused by how this assert fires: it shouldn't be possible when we have single-block regions.
Do you have an example IR?

Reply:
I'll get back to you soon. The usage of this function seems a bit non-trivial in this case, especially with block arguments.

Reply:
If you need examples for this API, there are a few in mlir/lib/Dialect/SCF/SCF.cpp and in mlir/test/lib/Dialect/Test/TestPatterns.cpp upstream.

Reply:
Yes, thanks. I was going through them.

Reply:
This is done.